Fix interoperation between DALI and PyTorch lightning due to buffering

JanuszL opened the pull request:
- PyTorch Lightning does double buffering - https://github.com/PyTorchLightning/pytorch-lightning/issues/1316
  and as DALI iterator owns the output tensors its content is trashed
  due to buffering. This PR improves the pytorch-lightning example
  by adding a copy of DALI outputs so they are no longer trashed

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

<!---
Thank you for contributing to NVIDIA DALI.

Please fill the relevant information in this PR template.

Fields with `- [ ]` represent check-boxes that can be marked after you create and save the Pull Request.
--->

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
<!---
Please explain what kind of change is in this PR and why it is submitted.
Any additional context or description of the solution is welcomed.
Examples:
- It fixes a bug *bug description*
- It adds new feature needed because of *why we need this feature*
- Refactoring to improve *what*
- The *new feature/bugfix* uses *a new data structure/algorithm* to do *X* instead of *Y*.
--->

#### Additional information
- Affected modules and functionalities:
<!--- Describe here what was changed, added, removed. --->

- Key points relevant for the review:
<!--- Describe here what is the most important part that reviewers should focus on. --->

## Checklist

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

### DALI team only

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A
<!---  Introduce new or affected requirement IDs, if applicable --->

**JIRA TASK**: N/A
<!--- DALI-XXXX or NA --->

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [2736473]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [2736473]: BUILD FAILED

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [2739793]: BUILD STARTED

klecki commented:
Two questions:
* at which point we create a dedicated PytorchLightning iterator?
* this code is now executed not on GPU. (At least it is printed as not using GPU in the logs). Is this intended?

JanuszL commented:
> Two questions:
> 
> * at which point we create a dedicated PytorchLightning iterator?

We need that in every case now. The first one just does the copy, the second does the additional processing. I have updated the text.

> * this code is now executed not on GPU. (At least it is printed as not using GPU in the logs). Is this intended?

Fixed. It seems that `Trainer` API has changed and I missed that.

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [2740134]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [2739793]: BUILD PASSED

dali-automaton commented:
**CI MESSAGE**: [2740134]: BUILD PASSED
